src/ai/backend/agent/server.py

```diff
@@ -591,26 +638,49 @@ async def __aexit__(self, *exc_info) -> None:
         await asyncio.sleep(0)
         if not self.debug_server_task.done():
             await self.debug_server_task
-        await self.agent.shutdown(self._stop_signal)
+        for agent in self.agents.values():
+            await agent.shutdown(self._stop_signal)
+        if self.metadata_server is not None:
+            await self.metadata_server.cleanup()
         await self.stats_monitor.cleanup()
         await self.error_monitor.cleanup()
 
+    def get_agent(self, agent_id: AgentId | None) -> AbstractAgent:
+        if agent_id is None:
+            agent_id = self._default_agent_id
+        if agent_id not in self.agents:
+            raise ValueError(
+                f"Agent '{agent_id}' not found in this runtime. "
+                f"Available agents: {', '.join(self.agents.keys())}"
+            )
+        return self.agents[agent_id]
+
     @collect_error
-    async def update_status(self, status):
-        await self.agent.etcd.put("", status, scope=ConfigScopes.NODE)
+    async def update_status(self, status, agent_id: AgentId):
+        agent_etcd = self.etcd_client_registry.get_client(agent_id)
+        await agent_etcd.put("", status, scope=ConfigScopes.NODE)
 
     @rpc_function
     @collect_error
-    async def update_scaling_group(self, scaling_group):
+    async def update_scaling_group(self, scaling_group: str, agent_id: AgentId | None = None):
+        agent = self.get_agent(agent_id)
+
         cfg_src_path = config.find_config_file("agent")
         with open(cfg_src_path, "r") as f:
             data = tomlkit.load(f)
-            data["agent"]["scaling-group"] = scaling_group
+        agent_configs = data.get("agents", [])
+        if agent_configs:
+            for agent_config in agent_configs:
+                if agent_config["agent"]["id"] == str(agent.id):  # type: ignore[index]
+                    agent_config["agent"]["scaling-group"] = scaling_group  # type: ignore[index]
+                    break
+        else:
+            data["agent"]["scaling-group"] = scaling_group  # type: ignore[index]
         shutil.copy(cfg_src_path, f"{cfg_src_path}.bak")
         with open(cfg_src_path, "w") as f:
             tomlkit.dump(data, f)
 
-        self.agent.update_scaling_group(scaling_group)
+        agent.update_scaling_group(scaling_group)
         log.info("rpc::update_scaling_group()")
```

Comment:
Let's handle the method separately for when agent_id is specified and when it is not specified.